src/main.rs

```diff
@@ -0,0 +1,296 @@
+    if data.len() != 1 || !data.contains_key("options") {
+        let keys: Vec<String> = data.keys().map(|k| k.to_string()).collect();
+        return Err(AppError::Validation(format!(
+            "Invalid YAML structure in {}: expected one top level group named 'options', found {:?}",
+            path, keys
+        )));
+    }
+
+    let options = data.get("options").expect("key to be guaranteed above");
+
+    // options should be a Mapping
+    if !options.is_mapping() {
+        return Err(AppError::Validation(format!(
+            "Invalid YAML structure in {}: expected 'options' to be a mapping",
+            path
+        )));
+    }
+
+    for (option, option_value) in options
+        .as_mapping()
+        .expect("options guaranteed to be a map above")
+    {
+        // TODO: verify option exists in schema
+        // if option not in schema[namespace]
+
+        // TODO: verify option value matches schema
+        // if option.type == schema[namespace][target][option].type
+
+        // Convert from serde_yaml::Value to serde_json::Value
+        let json_value = serde_json::to_value(option_value)?;
+
+        result.insert(
+            option.as_str().expect("option key to be valid").to_string(),
+            json_value,
+        );
+    }
+
+    Ok(result)
+}
+
+/// Checks options in the same target for duplicate keys
+fn ensure_no_duplicate_keys(grouped: &NamespaceMap) -> Result<()> {
+    for (_, targets) in grouped {
+        for (_, filedata) in targets {
+            let mut key_to_file = HashMap::<String, String>::new();
+            for FileData { path, data } in filedata {
+                for key in data.keys() {
+                    if let Some(first_file) = key_to_file.get(key) {
+                        return Err(AppError::DuplicateKey {
+                            key: key.to_string(),
+                            first_file: first_file.to_string(),
+                            second_file: path.to_string(),
+                        });
+                    }
+                    key_to_file.insert(key.to_string(), path.to_string());
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Merges keys from many option files into one map
+fn merge_keys(filedata: &[FileData]) -> OptionsMap {
+    let mut result = HashMap::new();
+    for FileData { data, .. } in filedata {
+        for (key, value) in data {
+            result.insert(key.clone(), value.clone());
+        }
+    }
+    result
+}
+
+/// Generates the list of output JSON files
+/// Uses the default target and handles overrides from other targets
+fn generate_json(maps: NamespaceMap) -> Result<Vec<(String, String)>> {
+    let mut json_outputs: Vec<(String, String)> = Vec::new();
+
+    // merge files in the same target together
+    for (namespace, targets) in maps {
+        let default_target = targets.get("default").ok_or_else(|| {
+            AppError::Validation(format!(
```

Comment:
It feels like we shouldn't need to wait until we're generating json to know that default target is missing. Assuming we're going to need this logic on PR's which doesn't necessarily need to generate the json